Add kep to move kubeproxy out of tree (to staging) (#2635)
kubernetes/enhancements: keps/sig-network/2634-kubeproxy-staging/README.md

## Summary
kube-proxy packages actually are present inside Kubernetes repo, inside 
[pkg/proxy](https://github.com/kubernetes/kubernetes/tree/v1.21.0/pkg/proxy) and 
the idea is to move it its own repo, already existing in 
[kubernetes/kube-proxy](https://github.com/kubernetes/kube-proxy) repo

## Motivation

We are seeing the rise of a lot of kube-proxy implementations. Although each of them 
follow their own dataplane implementation (IPTables, IPVS, eBPF, open flow programming), 
every implementation needs to follow the same logic: watching Services, Endpoints and 
most recently EndpointSlices.

Everytime we change the logic from Services, each of those implementations need to make 
changes into their code base to be compliant with the new features. Some of them vendor
the whole kubernetes/kubernetes repo, while others copy all the code base to their own
repository.

Moving pkg/proxy logics to its own repo will allow a better decoupling of the code, 
the evolution of proxy features without being tighten to kubernetes main code and also 
to future kube-proxy implementations to import this logic without needing to import the 
whole Kubernetes code.

> This was also motivated by [kubernetes #92369](https://github.com/kubernetes/kubernetes/issues/92369)

Author: No opposition to that (even if I had, Jay would kill me!) and also I have the feeling that this enhancement would be much more a buried pipe change than a real need. 

Originally this was created to make it easier for the CNI implementations but I guess now every one of the majors one already have their own logics/copy/etc so does this still makes sense?

### Goals

* Move kube-proxy repo to `kubernetes/kube-proxy` repo using staging.
* Decouple kube-proxy from internal kubernetes dependencies.

### Non-Goals
* Implementing interfaces for kube-proxy new implementations (like virtual-kubelet, 
but for proxies). This can be addressed on a later KEP

## Proposal

> kube-proxy already have its own repo. The proposal is to:

Reviewer: nit:
```suggestion
kube-proxy already has its own repository. The proposal is to:
```

### Move the internal network dependencies that kube-proxy already has

> kube-proxy have some internal dependencies, mostly notable in `kubernetes/pkg/utils`. We

Reviewer 2: We need a really really good test suite to make sure these dependencies which we move out can be tested independently of kubernetes. This is the same problem we had with sig-storage's try to move mount utils outside of k/k and eventually we had to bring it back into staging

Reviewer 3: I think that the proposal is the same as mount utils, create a new staging repo `staging/src/k8s.io/network-utils`

Reviewer 2: @aojea line 87 is a section that says `UNRESOLVED repository for network utils` :)

Author: Thanks @dims 

@jayunit100 @aojea you are my reference for testing things, can use some help here! :)

Author: Yes, so here's my point. We need to define where all those things needs to go. My proposal is to move to network-utils, BUT as we need to create a kep for moving kube-proxy to staging, I want to use this to discuss as well where everything should be.

async is the hardest example: @danwinship opened that PR in utils, but by utils guideline async should not be there (because it vendors k8s.io/client-go) so, maybe component-helpers?

We can use this Kep to justify why are we moving things the way we are :)

Reviewer 4: > async is the hardest example: @danwinship opened that PR in utils, but by utils guideline async should not be there (because it vendors k8s.io/client-go) so, maybe component-helpers?

The PR (https://github.com/kubernetes/utils/pull/165) also moves the relevant parts of `"k8s.io/client-go/util/flowcontrol"` into utils as well.

But anyway, I think I originally filed that PR before component-helpers existed (or at least before I'd heard of it), and maybe component-helpers would be a better place

Reviewer 5: Yes we are working on better tests for kube proxy at the moment --- @dougsland is building some prototypes well have an update by the end of the wk

Reviewer 3: 1. out of tree  :/ 
https://github.com/kubernetes/enhancements/pull/2635/files#r615445820 and  https://github.com/kubernetes/enhancements/pull/2635/files#r615453858

2. component-helpers, i.e. does kubectl needs to vendor moby/ipvs?   that means that a CVE in ipvs will flah kubectl ... do

3. if we move to `staging/src/k8s.io/net-utils` , we do the same that storage with mount-utils and we don't create undesired dependencies

am I missing something? is not 3 the best alternative?

Author: I'm a huge +1 to move the networking utils to k/net-utils

Reviewer 6: I know we KIND OF want to avoid polluting the "k8s.io" space (e.g. "k8s.io/iptables", "k8s.io/ipvs", "k8s.io/conntrack" all feel somewhat "big".  Additionally, we probably don't want to be telling the whole go ecosystem "hey, use this library" because, frankly, we aren't putting that much thought into generalizing theme.

Some alternatives to an omnibus "util" repo.

1) Make lots of small repos in `util.k8s.io/...` - we can still keep the source of truth in k/k (the monorepo) and publish to these if we need.  "util.k8s.io/iptables" feels less "endorsed".

2) Make lots of small repos in `internal.k8s.io` - same as above but even more "warning".

3) Adopt a convention that makes these libs feel less "big".  E.g. "libiptables" or something?

Reviewer 4: > Make lots of small repos in internal.k8s.io - same as above but even more "warning".

That seems plausible...

> Adopt a convention that makes these libs feel less "big"

I don't understand your comments about big-ness... Why do we care whether they _feel_ big as opposed to caring whether they _are_ big?

Reviewer 6: This topic is on the agenda for sig-arch code organization next week  (I just missed last week's)

some still vendors external dependencies (like `moby/ipvs`).
If we move every pkg/utils used by kube-proxy to component helpers, another project
already vendoring it (like kubectl) might vendor moby/ipvs because of transitive 
dependencies.
We need to define if network utils should be moved to its own repo (suggesting kubernetes/net-utils)
or if we can accept that other projects vendoring `component-helpers` are going to vendor the 
dependencies of kube-proxy as well
<<[/UNRESOLVED]>>

### Move other dependencies kube-proxy already has inside k/k
kube-proxy vendors some packages that are used by other parts of kubernetes codebase. 
From now, we can point:
* pkg/util/async (async bounder) which is used by pkg/controlplane
* pkg/util/sysctl which is used by pkg/kubelet and kubemark

### Move feature gate package dependency
Because kube-proxy relies on defined Feature gates, it vendors `pkg/features`.

This needs to be copied/moved the same way controller manager already does, moving this to
`kubernetes/kube-proxy/pkg/features` and changing on the code, and putting all the feature
gates referenced by kube-proxy there

> ### Move pkg/proxy to staging

Reviewer 6: Should we also do cmd/kube-proxy?

Author: should we? not sure, I think that right now cmd/kube-proxy is in a good place where it is (following the same approach of kubelet/scheduler/etc) that are there.

But you are here on a really longer time that I am, so I may be missing something really important!!

Reviewer 6: IMO the split between cmd/kube-proxy and pkg/proxy is pretty much completely artificial.  Nothing uses pkg/proxy except cmd/proxy (as far as I caan tell).  Repeat for kubelet, controllers, etc and their respective binaries.  It just makes the code harder to find.

That said, this is FAR from the most important thing going on right now, so I don't know if it makes sense to do a bunch of busywork?

The final step is to move pkg/proxy to `staging/src/k8s.io/kube-proxy/pkg` and let people willing
to vendor kube-proxy to change from kubernetes/kubernetes to `kubernetes/kube-proxy` repo.

### Risks and Mitigations

> "If a project vendors Kubernetes to import kubeproxy code, this will break them. 
> On the bright side, afterwards, these importers will have a much cleaner path to 
> include kubeproxy code. Before moving forward with this plan, we will identify 
> and communicate to these projects." - Shameless copied from 

Reviewer: nit:
```suggestion
> If a project vendors Kubernetes to import kubeproxy code, this will break them. 
> On the bright side, afterwards, these importers will have a much cleaner path to 
> include kubeproxy code. Before moving forward with this plan, we will identify 
> and communicate to these projects.

&ndash;Shamelessly copied from 
```
